### PR TITLE
Start making the rollout groups parallel

### PR DESCRIPTION
**Pull request by metalmatze**
This might not be perfect yet. I tried it with one example, but there may be some things to be improved.

### CODE REVIEW

File: rollout/rollout.go [hunk 1]
```diff
@@ -122,14 +123,28 @@ func (r *Runner) Execute(rolloutConfig *Config) (err error) {
 	}
 
 	for _, group := range res.Rollout.Spec.Groups {
+		var wg sync.WaitGroup
+		errs := make(chan error, len(group.Steps))
```
Comment (brancz) [inline rollout/rollout.go:127]: I'd say let's rather just make this a multi-error that is protected by a mutex.

**Review by brancz — COMMENTED** [review]
Neat!

File: rollout/rollout.go [hunk 2]
```diff
@@ -122,14 +123,28 @@ func (r *Runner) Execute(rolloutConfig *Config) (err error) {
...
+					errs <- fmt.Errorf("could not find object named %q", step.Object)
+					return
+				}
+				if err := r.runStep(step, object); err != nil {
+					errs <- err
+					return
+				}
+			}(step)
+
+			wg.Wait()
```
Comment (kakkoyun) [inline rollout/rollout.go:145]: Maybe I'm reading this incorrectly but shouldn't this wait have to be one upper scope? (outer for)

File: rollout/rollout.go [hunk 3]
```diff
@@ -122,14 +123,28 @@ func (r *Runner) Execute(rolloutConfig *Config) (err error) {
...
+
+			if err := <-errs; err != nil {
```
Comment (kakkoyun) [inline rollout/rollout.go:147]: I don't think `errs` channel ever gets closed.

**Review by kakkoyun — COMMENTED** [review]
Looking great. I've pointed some concerns about synchronization.

File: rollout/rollout.go [hunk 4]
```diff
@@ -4,19 +4,21 @@ import (
...
 	"time"
 
+	"github.com/brancz/locutus/client"
+	"github.com/brancz/locutus/feedback"
 	"github.com/brancz/locutus/render"
+	"github.com/brancz/locutus/rollout/checks"
+	"github.com/brancz/locutus/rollout/types"
 	"github.com/go-kit/kit/log"
+	"github.com/hashicorp/go-multierror"
+	_ "github.com/hashicorp/go-multierror"
```
Comment (brancz) [inline rollout/rollout.go:17]: What's this about? :thinking:

File: rollout/rollout.go (hunk 2 above)
Comment (metalmatze) [reply rollout/rollout.go:151]: Wow. So right. Oops.

**Comment by metalmatze** [comment]
Properly updated this PR now. I think I didn't even push a few changes I had already done :innocent:

**Review by brancz — APPROVED** [review]
lgtm but giving @kakkoyun another chance

**Review by kakkoyun — APPROVED** [review]
🥇LGTM

**Pull request merged**